### PR TITLE
Add Switch Transformer implementation

### PR DESCRIPTION
**Pull request by ksalama**
Hi @fchollet and @apoorvnandan -

I extended the Text Classification with Transformer to add [Switch Transformer](https://arxiv.org/abs/2101.03961) implementation. I restructured the example to allow running two experiments: 1) with the standard Transformer block, and 2) with the Switch Transformer. In addition, I changed the Transformer block implementation to allow either 1) having a standard feedforward network, or 2) having a Switch layer.

I was thinking of creating a separate example, but I though it would be better to have a single example with both implementations. Happy to move it to a separate one.

Let me know what do you think.

Cheers

### CODE REVIEW

**Comment by fchollet** [comment]
Thanks for the PR!

> I was thinking of creating a separate example, but I though it would be better to have a single example with both implementations. Happy to move it to a separate one.

I think "text classification with Transformer" and "text classification with Switch Transformer" are different enough that it makes sense to have 2 separate examples instead of just one. This will also allow us to keep the original example as short and simple as possible.

**Comment by ksalama** [comment]
@fchollet - I have split the examples now. Cheers

File: examples/nlp/text_classification_with_switch_transformer.py [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+        y_train,
+        batch_size=batch_size,
+        epochs=num_epochs,
+        validation_data=(x_val, y_val),
+    )
+    return history
+
+
+classifier = create_classifier()
+run_experiment(classifier)
```
Comment (fchollet) [inline examples/nlp/text_classification_with_switch_transformer.py:322]: We'd need some final display of test results and comment about these results

File: examples/nlp/text_classification_with_switch_transformer.py [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+
+
+class Router(layers.Layer):
+    def __init__(self, num_experts, expert_capacity):
+        self.num_experts = num_experts
+        self.route = layers.Dense(units=num_experts)
+        self.expert_capacity = expert_capacity
+        super(Router, self).__init__()
+
+    def call(self, inputs, training):
```
Comment (fchollet) [inline examples/nlp/text_classification_with_switch_transformer.py:142]: Set a default value for `training`, e.g. `training=False`

**Review by fchollet — COMMENTED** [review]
Thanks for the PR. Looking good! 👍

File: examples/nlp/text_classification_with_switch_transformer.py (hunk 2 above)
Comment (ksalama) [reply examples/nlp/text_classification_with_switch_transformer.py:142]: Done

File: examples/nlp/text_classification_with_switch_transformer.py (hunk 1 above)
Comment (ksalama) [reply examples/nlp/text_classification_with_switch_transformer.py:322]: - We provide a validation data in the `fit` method , for which accuracy is displayed during training. I didn't add the final display for this split as you suggested in the other example.
- I added a "general" conclusion text, but not commenting on the produced results as we are not comparing it with the standard Transformer. 

What do you think?

**Comment by ksalama** [comment]
@fchollet - Thanks a lot for your comments. I addressed them

Let me know if you have any further ones

**Review by fchollet — COMMENTED** [review]
Thanks for the update! LGTM, please add the generated files.

**Comment by ksalama** [comment]
@fchollet - Thanks for your review. I have added the generated files.

**Comment by fchollet** [comment]
Awesome, thank you! 👍

**Review by fchollet — APPROVED** [review]
